src/ast/mod.rs

```diff
@@ -1048,45 +1263,45 @@ where
     Leaf: Clone + Eq + Display + Hash,
     Ann: Clone + Default,
 {
+    /// Applys `f` to all sub-expressions (including `self`)
+    /// bottom-up, collecting the results.
     pub fn map_exprs(
-        &self,
-        f: &mut impl FnMut(&GenericExpr<Head, Leaf, Ann>) -> GenericExpr<Head, Leaf, Ann>,
+        self,
+        f: &mut impl FnMut(GenericExpr<Head, Leaf, Ann>) -> GenericExpr<Head, Leaf, Ann>,
     ) -> Self {
         match self {
             GenericAction::Let(ann, lhs, rhs) => {
-                GenericAction::Let(ann.clone(), lhs.clone(), f(rhs))
+                GenericAction::Let(ann.clone(), lhs.clone(), rhs.map(f))
             }
+            // TODO should we refactor `Set` so that we can map over Expr::Call(lhs, args)?
```

Comment:
I like `Set` to have an `lhs` and `args` more so that the constraints that the first argument to `Set` is not a literal nor variable is enforced. It is just a good practice to make illegal states unrepresentable. Here, we essentially ensure that the substituted result produced by `f` is not a literal/variable.

Reply:
keeping as is for now